Registry - corrupted tags handling

bartebor opened the pull request:
This PR fixes https://github.com/docker/distribution/issues/2430.
As I have received no feedback I decided to create MR with my proposal.

There are two modifications:

1. When tagging, if creating `current/link` fails, we try to remove tag to avoid storage corruption and return error as before.

2. When deleting manifest, we allow to remove corrupted tags with no `current/link` if there is matching blob link. Other tags do not break tag scanning and are reported in logs.
Corrupted tags can be removed via API by digest if it is known. It could also be done with some asynchronous GC-like mechanism in the future.

stevvooe commented:
How can this code detect whether the link was created by the caller or it was due to racing with another creation attempt? As soon as you detect something is deletable, it may no longer be deletable due to another concurrent operation.

I suspect the source of the issue here is swift's consistency model.

What are the actual errors you are receiving?
What is an example flow of push and pull that is encountering this error?

bartebor commented:
> How can this code detect whether the link was created by the caller or it was due to racing with another creation attempt? As soon as you detect something is deletable, it may no longer be deletable due to another concurrent operation.

I am not sure which case you are talking about, so I address both.
When tagging (PUT manifest), `current/link` is created only once. If tag were to be moved, new `current/link` replaces old link. I believe this is done atomically, at least on registry's end. There are no deletions of this file other than untagging, when whole tag directory is removed. The latter (at least in swift) is not atomic. There is teoretical possibility of interference between two concurrent PUTs to the same tag, but it is not possible in my environment.
Either way, PUT is not atomic and can fail half way leaving corrupted storage. My solution is not bullet proof, because untag can also fail, hence changes in lookup loop.

When untagging (DELETE manifest) we can find tag with no `current/link`. In 2.6.2 this terminates loop and makes it impossible to delete manifests. I changed this behavior to ignore such tag and continue.
Howewer, if I find matching link in such tag, I mark this tag for deletion. There is possibility that this tag has more than one link (it is some previously-moved tag, like `latest`) and that our digest was not latest. It makes no difference though, because this tag is unusable anyway.

> I suspect the source of the issue here is swift's consistency model.

Yes, (at least) swift driver has non-atomic operations, which may fail somewhere in the middle of processing causing strange behaviours. Delete is implemented recursively, Walk fetches list of objects in given path and then enumerates them while those can be changed/deleted... 
But as I described PUT manifest behaviour, corruptions happen not only because of swift driver design. 

> What are the actual errors you are receiving?

Actual error is HTTP 500 when deleting any manifest after corrupted tag appeared in registry.
Logs look like that:
```
dockerhub time="2018-01-02T23:10:00.41700727Z" level=debug msg="swift.GetContent(\"/docker/registry/v2/blobs/sha256/e4/e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c/data\")" go.version=go1.7.3 http.request.host="dockerhub:443" http.request.id=73f5fe84-3ca7-4960-aa81-43fb66a7ca60 http.request.method=GET http.request.remoteaddr=10.32.216.138 http.request.uri="/v2/szwadron/overseer-webserver/manifests/communication_c2398eeb-66e2-4af5-8bee-7eecf2a894a7" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" instance.id=677ccbe3-3260-4e76-83cf-2dfb8d07bbdb service=registry trace.duration=7.980071ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=3d4d83c1-dacb-4a57-806d-6becd49680d0 trace.line=82 vars.name="szwadron/overseer-webserver" vars.reference="communication_c2398eeb-66e2-4af5-8bee-7eecf2a894a7" version=v2.6.1 
...
dockerhub time="2018-01-02T23:10:21.23563118Z" level=debug msg="swift.GetContent(\"/docker/registry/v2/repositories/szwadron/overseer-webserver/_manifests/tags/communication_4a4e725c-d96f-4623-8876-6ac2bd36967b/current/link\")" go.version=go1.7.3 http.request.host="dockerhub:443" http.request.id=31350391-91b8-4c35-9de9-d7deae5dc79b http.request.method=DELETE http.request.remoteaddr=10.32.216.138 http.request.uri="/v2/szwadron/overseer-webserver/manifests/sha256:e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" instance.id=677ccbe3-3260-4e76-83cf-2dfb8d07bbdb service=registry trace.duration=7.451283ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=6bfbe2e1-1a65-48ca-ade1-7e96fa18ea2a trace.line=82 vars.name="szwadron/overseer-webserver" vars.reference="sha256:e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c" version=v2.6.1 
dockerhub time="2018-01-02T23:10:21.235893239Z" level=error msg="response completed with error" err.code=unknown err.message="swift: Path not found: /docker/registry/v2/repositories/szwadron/overseer-webserver/_manifests/tags/communication_4a4e725c-d96f-4623-8876-6ac2bd36967b/current/link" go.version=go1.7.3 http.request.host="dockerhub:443" http.request.id=31350391-91b8-4c35-9de9-d7deae5dc79b http.request.method=DELETE http.request.remoteaddr=10.32.216.138 http.request.uri="/v2/szwadron/overseer-webserver/manifests/sha256:e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" http.response.contenttype="application/json; charset=utf-8" http.response.duration=20.811199051s http.response.status=500 http.response.written=240 instance.id=677ccbe3-3260-4e76-83cf-2dfb8d07bbdb service=registry vars.name="szwadron/overseer-webserver" vars.reference="sha256:e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c" version=v2.6.1 
dockerhub 127.0.0.1 - - [02/Jan/2018:23:10:00 +0000] "DELETE /v2/szwadron/overseer-webserver/manifests/sha256:e428b2acd7e0af474d9f5273ffde538d65a5cb65136aeea1072efef2bcf1041c HTTP/1.0" 500 240 "" "Apache-HttpClient/4.3.6 (java 1.5)"
```
Unfortunately I have no logs from failed PUT, but it was the case when creating (first) link succeded, and creating `current/link` failed. PUT failed with 5xx so client software tried again some time later.

> What is an example flow of push and pull that is encountering this error?

I can not give you examples of `docker` command, as we are using V2 API directly, but I will try to describe our use case.
We have some docker image (overseer-webserver) which is unchanged most of the time (we use constant version tag 0.0.10).
Many times a day a tests are performed. From docker registry perspective there are following actions:
- create new unique tag on test image: GET manifest from 0.0.10 tag; alter manifest to make it unique; PUT altered manifest at a unique tag. Altering manifest is needed, because there is no way to delete tag by name. When deleting by digest, all tags with that digest is deleted.
- perform testing (deploy tagged image in kubernetes cluster, perform tests, undeploy)
- GET manifest from unique tag created earlier to obtain its digest (this fails if we try to fetch corrupted tag)
- DELETE manifest from digest (this fails if we try to delete any tag while any of them is corrupted)

We need to create a new, constant tag to avoid possibility of tag mutation during deployment.

The tests execute concurrently, so there is possibility of undesired interactions (in fact I see them when two DELETEs run at the same moment, but those are resolved eventually). Normally there should be about 100 docker tags max, but from time to time some of them leak and tag count grows (I will investigate this).

bartebor commented:
Sorry for closing and reopening, I pushed wrong button after writing so long post :wink: .

stevvooe commented:
> There is teoretical possibility of interference between two concurrent PUTs to the same tag, but it is not possible in my environment.

Unfortunately, this is not true for everyone's environment and affects whether or not we could take such a patch. You can read about some of the issues that may be seen in https://github.com/docker/distribution/blob/master/docs/architecture.md#eventual-consistency. If you can adjust swift's consistency model (not sure if this is possible), the problem will go away. Basically, the issue is that swift's visibility after an operation makes so that a corrupt repo layout is written. We need to figure out which operation is failing and causing that corruption before deleting data.

The main issue with this patch is that this relies on a race condition to be correct:

```
if deletable() {
  // ... deletable can be false here after reading
} else {
  // ... deletable may be true here after reading
}
```

Because `deletable()` may change as soon as it is read, it is basically random which block is actually taken at runtime. The potential for corruption is _even worse_. To fix this issue, you need to address the problems in the storage backend that prevent the correct writes from happening in the first place.

From this point, I think it is best if you could focus on identifying the problem that leads to the corruption before finding a solution. This would be a description of what _exactly_ is happening, such as which path is removed and which errors are encountered during that removal process. This needs the following:

1. Unedited logs from the first corruption point.
2. A trace of requests from your client and the source code, if possible.
3. A layout (ie tree or something similar) of the repository before and after the corruption.

If you have a reproduction with the filesystem driver, that would go even further to isolate whether or not this is a problem with swift or the registry code.

The other possibility here is to redefine a workflow that doesn't so heavily rely on tag deletions. Just define a unique tag for each artifact you push with a version number and don't change it. When that gets pushed, update your deployment and push the tag.

bartebor commented:
I understand your concerns of concurrency and I agree. The problem is, that in some situations repository is rendered unusable, which is simply not acceptable in prodution environments. If we want to use remote storage at all, we should have a code that does not corrupt or tolerates corruptions on that remote storage or eventually fixes those.
In this very case the problem lies in this function from `storage/tagstore.go` (you may treat it as identification of source of problem):
```
func (ts *tagStore) Tag(ctx context.Context, tag string, desc distribution.Descriptor) error {
	currentPath, err := pathFor(manifestTagCurrentPathSpec{
		name: ts.repository.Named().Name(),
		tag:  tag,
	})

	if err != nil {
		return err
	}

	lbs := ts.linkedBlobStore(ctx, tag)

	// Link into the index
	if err := lbs.linkBlob(ctx, desc); err != nil {
		return err
	}

	// Overwrite the current link
	return ts.blobStore.link(ctx, currentPath, desc.Digest) // suppose this fails....
}
```
This has nothing to do with storage driver, any storage operation can fail anytime. This function is not atomic - if last statement in this function fails for whatever reason (network error, etc.), we end up with corrupted repository. GC does not handle this, manifest delete does not handle this and fetching manifest by tag name does not handle it neither. My solution can be too far going because of mutation of repository state (which can fail too) and I understand this might cause problem in some setups. We could leave this code as it is and try to mitigate corruption in other places.

Ok, let's go to second (manifest deletion) part.
Current implementation returns error when any tag is corrupted by missing `current/link` file. This means that no tags can be deleted, so their number grows unbounded. We reach thousands of tags, repository gets terribly slow and reaches point that it becomes unusable for all repos, not only the one corrupted.

Just like with PUT, I added deletion of such tags and just as before this is not ideal solution because of concurrent execution and inevitable errors accessing storage. So maybe we could skip deletion and just _ignore_ such broken tags? This could potentially keep tags being currently created from deletion, but this should be no problem; if client(s) not synchronize create/delete attempts they should be prepared for any execution order and handle it themselves. Even without deletion this solution is far better than current implementation for repository as a whole since it does not DOS it (at least so quickly). Do you consider this as an acceptable solution?

If we removed deletions then I think we _should_ have some GC-like sanitizer to repair inconsistencies.

I'm perfectly fine with idea that for _some_ time repository returns 500's for affected operations, because as a repository client I should be prepared for error and keep retrying until operation succeeds. This would mean that we reach desired repository state eventually.

I'm far from enforcing my patch as a best solution, I just wanted it to be a start of discussion on real problem and have your help in finding the best way to solve it. 

> From this point, I think it is best if you could focus on identifying the problem that leads to the corruption before finding a solution. This would be a description of what exactly is happening, such as which path is removed and which errors are encountered during that removal process. This needs the following:
> 1. Unedited logs from the first corruption point.
> 2. A trace of requests from your client and the source code, if possible.
> 3. A layout (ie tree or something similar) of the repository before and after the corruption.

I think I made clear what the root cause of problem was. Believe me, I'm often in need of diagnostic information too and I know how it feels not having it :smile:  
I do not have logs from the moment of failure because of log retention - the logs have been deleted. This situation is rare; it happened only twice. I digged the logs when it happened second time and found what I've written in https://github.com/docker/distribution/issues/2430.

> If you have a reproduction with the filesystem driver, that would go even further to isolate whether or not this is a problem with swift or the registry code.

Some class of errors just not happen locally. While preparing this patch I was deliberately corrupting repository to test delete manifest logic, but was unable to make tag operation to fail in a way observed in production. I think I would have to make some remote storage mock and simulate this with it. If I had production repository on filesystem driver instead of swift I would probably have no problems to report...

> The other possibility here is to redefine a workflow that doesn't so heavily rely on tag deletions. Just define a unique tag for each artifact you push with a version number and don't change it. When that gets pushed, update your deployment and push the tag.

This is not possible. We _could_ fix my testing environment like this, but it would be not practical to enforce such practices on our clients (we are developing some sort of CaaS system on kubernetes; some clients tend to use such constant tags, but we can not take it for granted - we need to create a tag to guarantee stability of deployment and potential rollback from earlier versions so we designed our functional test suite to use worst-case scenario). We garbage collect deployments eventually and this is the moment when above problem huts us. It is worth saying that, unlike our clients' deployments, our test really stresses test repository but hey - that is how we got aware of this problem. Anyway - we all want hub to just work, so let's make it so, I'll be glad to help make it happen.

stevvooe commented:
@bartebor I'm unconvinced this is not a problem with swift or the swift driver. We haven't seen this behavior in other drivers (that I know of).

Is there a solution that will simply get the right result on the next push? For example, if the link operation fails, it should just link it on the next call, without deleting anything.

Again, I need to see the layout before and after the corruption to understand what is going on. If you could even just write it out as what you think is happening, that would be helpful.

bartebor commented:
This is not swift driver's fault, because two file creations in `Tag` are not transactional and may end up partially executed. It is just much more likely that this happens on remote storage and lossy network opposing to local filesystem/rock solid network. The fact you have not hed such issue reported may be also that my usage scenario is susceptible to a such an fault resulting in DoS. Normally one would just push image again to have problem fixed. But if it is not the very same tag pushed (the same manifest digest to be exact), i.e CI job incremented build number, user may not notice that they lost ability to delete tags.

I have prepared a simulation which you can run yourself. Assume running a local repository at `devhub:5100` (`devhub` is a name in `etc/hosts` pointing to localhost):
```
$ cat config-devhub.yml
version: 0.1
log:
  level: debug
  fields:
    service: registry
    environment: development
storage:
    delete:
      enabled: true
    filesystem:
        rootdirectory: /tmp/docker-registry
    maintenance:
        uploadpurging:
            enabled: false
http:
    addr: :5100
    debug:
        addr: localhost:5101
    headers:
        X-Content-Type-Options: [nosniff]
health:
  storagedriver:
    enabled: true
    interval: 10s
    threshold: 3
$ go run main.go serve config-devhub.yml
INFO[0000] debug server listening localhost:5101        
WARN[0000] No HTTP secret provided - generated random secret. This may cause problems with uploads if multiple registries are behind a load-balancer. To provide a shared secret, fill in http.secret in the configuration file or set the REGISTRY_HTTP_SECRET environment variable.  environment=development go.version=go1.9.2 instance.id=9b317ce5-35e8-44c4-933c-87e5eab8185d service=registry version=v2.6.2+unknown
INFO[0000] redis not configured                          environment=development go.version=go1.9.2 instance.id=9b317ce5-35e8-44c4-933c-87e5eab8185d service=registry version=v2.6.2+unknown
INFO[0000] listening on [::]:5100                        environment=development go.version=go1.9.2 instance.id=9b317ce5-35e8-44c4-933c-87e5eab8185d service=registry version=v2.6.2+unknown
...
```
Now, create and push two simple docker images:
```
$ touch file1 file2
$ cat > Dockerfile1
FROM scratch
ADD ./file1 /
$ cat > Dockerfile2
FROM scratch
ADD ./file2 /
$ docker build -t devhub:5100/test:test1 -f Dockerfile1 .
Sending build context to Docker daemon  22.88MB
Step 1/2 : FROM scratch
 ---> 
Step 2/2 : ADD ./file1 /
 ---> Using cache
 ---> e6e9a41f112f
Successfully built e6e9a41f112f
$ docker build -t devhub:5100/test:test2 -f Dockerfile2 .
Sending build context to Docker daemon  22.88MB
Step 1/2 : FROM scratch
 ---> 
Step 2/2 : ADD ./file2 /
 ---> Using cache
 ---> da141f4945cd
Successfully built da141f4945cd
$ docker push devhub:5100/test:test1
The push refers to a repository [devhub:5100/test]
fb502e56d48d: Pushed 
test1: digest: sha256:a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e size: 523
$ docker push devhub:5100/test:test2
The push refers to a repository [devhub:5100/test]
56382659a064: Pushed 
test2: digest: sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 size: 523
```
Up to now everything works fine. Repository layout:
```
$ cd /tmp/docker-registry/docker/registry/v2
$ tree
.
├── blobs
│   └── sha256
│       ├── a5
│       │   └── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
│       │       └── data
│       ├── ba
│       │   └── bae58460bb1ad7692e80ee9c21148d6a71562fa6976f5404c2b2d87de5f08f8c
│       │       └── data
│       ├── da
│       │   └── da141f4945cd8e4456caba79138f06deb313579a65940f6d97beb05fd3ccf81c
│       │       └── data
│       ├── e6
│       │   └── e6e9a41f112f39ce3a7c245fbe1a4b5048dbbd9bc35320559f70c6d4a3eca37f
│       │       └── data
│       ├── f7
│       │   └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
│       │       └── data
│       └── f8
│           └── f857fdc27c8e3dcfd13da8c60e7e522dda713c8125facedd69568466d32324ab
│               └── data
└── repositories
    └── test
        ├── _layers
        │   └── sha256
        │       ├── bae58460bb1ad7692e80ee9c21148d6a71562fa6976f5404c2b2d87de5f08f8c
        │       │   └── link
        │       ├── da141f4945cd8e4456caba79138f06deb313579a65940f6d97beb05fd3ccf81c
        │       │   └── link
        │       ├── e6e9a41f112f39ce3a7c245fbe1a4b5048dbbd9bc35320559f70c6d4a3eca37f
        │       │   └── link
        │       └── f857fdc27c8e3dcfd13da8c60e7e522dda713c8125facedd69568466d32324ab
        │           └── link
        ├── _manifests
        │   ├── revisions
        │   │   └── sha256
        │   │       ├── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
        │   │       │   └── link
        │   │       └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
        │   │           └── link
        │   └── tags
        │       ├── test1
        │       │   ├── current
        │       │   │   └── link
        │       │   └── index
        │       │       └── sha256
        │       │           └── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
        │       │               └── link
        │       └── test2
        │           ├── current
        │           │   └── link
        │           └── index
        │               └── sha256
        │                   └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
        │                       └── link
        └── _uploads
```
Let's now simulate that a second operation in `Tag` failed like in my case, and delete current link from first tag:
```
$ rm -r repositories/test/_manifests/tags/test1/current
$ tree repositories/test/_manifests/tags
repositories/test/_manifests/tags
├── test1
│   └── index
│       └── sha256
│           └── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
│               └── link
└── test2
    ├── current
    │   └── link
    └── index
        └── sha256
            └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
                └── link
```
Now we can try some tests:
```
$ docker pull devhub:5100/test:test1
Pulling repository devhub:5100/test
Error: image test:test1 not found
$ curl http://devhub:5100/v2/test/tags/list # note two tags are returned
{"name":"test","tags":["test2","test1"]}
$ curl -H'Accept: application/vnd.docker.distribution.manifest.v2+json' http://devhub:5100/v2/test/manifests/test1 # corrupted tag - we can not determine its digest, so we can not delete it
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Tag":"test1"}}]}
$ curl -H'Accept: application/vnd.docker.distribution.manifest.v2+json'  http://devhub:5100/v2/test/manifests/test2 # this works fine
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 1278,
      "digest": "sha256:da141f4945cd8e4456caba79138f06deb313579a65940f6d97beb05fd3ccf81c"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 93,
         "digest": "sha256:f857fdc27c8e3dcfd13da8c60e7e522dda713c8125facedd69568466d32324ab"
      }
   ]
}
$ curl -XDELETE http://devhub:5100/v2/test/manifests/sha256:a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e # we should not know tag1 digest, but even if we do know, delete fails
{"errors":[{"code":"UNKNOWN","message":"unknown error","detail":{"Path":"/docker/registry/v2/repositories/test/_manifests/tags/test1/current/link","DriverName":"filesystem"}}]}
$ curl -XDELETE http://devhub:5100/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 # deleting _not_corrupted_ tag also fails
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
```
Docker registry logs from last DELETE:
```
DEBU[0040] filesystem.Stat("/")                          environment=development go.version=go1.9.2 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=44.801µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat trace.id=2e4844ac-babc-45e3-9cca-a2b88c4e7a0c trace.line=137 version=v2.6.2+unknown
DEBU[0044] authorizing request                           environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] DeleteImageManifest                           environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] (*manifestStore).Delete                       environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.GetContent("/docker/registry/v2/repositories/test/_manifests/revisions/sha256/f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79/link")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=38.715µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent trace.id=b47ded86-58e3-4cbb-b1be-2a0a2949f5ff trace.line=82 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.Stat("/docker/registry/v2/blobs/sha256/f7/f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79/data")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=30.808µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat trace.id=58c2c910-7d1f-4995-b360-b82f34a2e468 trace.line=137 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.Delete("/docker/registry/v2/repositories/test/_manifests/revisions/sha256/f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79/link")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=67.37µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).Delete trace.id=1f248b33-33a0-4dd3-ac81-dfb6cd8f6724 trace.line=177 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.Delete("/docker/registry/v2/repositories/test/_layers/sha256/f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79/link")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=23.699µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).Delete trace.id=dfd208f9-528b-4ce7-abaa-21a20c5f7dfb trace.line=177 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.List("/docker/registry/v2/repositories/test/_manifests/tags")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=33.717µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).List trace.id=8a2cf8f3-1858-4a3b-bfc6-eed2d0cf351d trace.line=150 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.GetContent("/docker/registry/v2/repositories/test/_manifests/tags/test2/current/link")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=26.185µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent trace.id=62b3c5a3-f88a-4e85-b9a7-b7814a3e6b38 trace.line=82 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
DEBU[0044] filesystem.GetContent("/docker/registry/v2/repositories/test/_manifests/tags/test1/current/link")  environment=development go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry trace.duration=13.037µs trace.file=github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent trace.id=42777c41-a3ed-425f-9ffb-6d21cff1cbc6 trace.line=82 vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
ERRO[0044] response completed with error                 environment=development err.code=unknown err.message=filesystem: Path not found: /docker/registry/v2/repositories/test/_manifests/tags/test1/current/link go.version=go1.9.2 http.request.host=devhub:5100 http.request.id=b4390f79-9e6a-4990-b4b1-c343a5fee0cc http.request.method=DELETE http.request.remoteaddr=127.0.0.1:49492 http.request.uri=/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 http.request.useragent=curl/7.57.0 http.response.contenttype=application/json; charset=utf-8 http.response.duration=1.673252ms http.response.status=500 http.response.written=177 instance.id=552c1457-4985-4d30-acef-88142e367836 service=registry vars.name=test vars.reference=sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 version=v2.6.2+unknown
127.0.0.1 - - [05/Feb/2018:09:11:31 +0100] "DELETE /v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 HTTP/1.1" 500 177 "" "curl/7.57.0"
```
Note that last `DELETE` also executed partially making it impossible to get manifest and delete it properly. This has less impact howewer, because it does not prevent _other_ tags from removal like partial `Tag` does.
```
$ curl -n -H'Accept: application/vnd.docker.distribution.manifest.v2+json' http://devhub:5100/v2/test/manifests/test2
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Name":"test","Revision":"sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79"}}]}
$ tree repositories/test/_manifests
repositories/test/_manifests
├── revisions
│   └── sha256
│       ├── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
│       └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
└── tags
    ├── test1
    │   └── index
    │       └── sha256
    │           └── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
    │               └── link
    └── test2
        ├── current
        │   └── link
        └── index
            └── sha256
                └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
                    └── link
$ curl -H'Accept: application/vnd.docker.distribution.manifest.v2+json' http://devhub:5100/v2/test/manifests/test2
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Name":"test","Revision":"sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79"}}]}
```
As you see from now on repository corruption progresses and there is simply nothing we can do. We could push those images if we had them, but we do not. I have static images stored in the registry, which are only tagged and untagged with `PUT` and `DELETE` manifest API calls as those are the only operations I need. There are no images to push to fix repository state. I could theoretically push each time (I do not even have docker on machines coordinating our CaaS so I do not), but those would be with unique tags, so it would repair nothing. I'm using documented API after all, so I expect it to do The Right Thing :)

In a distributed system random errors are inevitable. That is why I would like repository to heal itself with no user intervention. 

I hope I was clear this time and I'm looking forward to have your opinion on what we can do to make registry more fault tolerant.

stevvooe commented:
@bartebor I think #2516 will handle this appropriately. We just need to ignore the second delete.

bartebor commented:
#2516 will not help in this case, because `DeleteImageManifest` function fails in `Lookup`, before we have a chance to call `Untag` you modified. Therefore we will delete manifest link, but leave tag links (`current`, `index`) behind with every subsequent DELETE. We need a fix in `Lookup`.

The root cause is failure in creating `<tag>/current/link` after successful creation of `<tag>/index/sha256/<digest>/link` when creating a new tag by `Tag`.

This patch should fix #2514 however. I encountered such error responses, but this was not a problem for me, because I issue retries until expected response is returned (`404 Not Found` in this case, meaning that a tag does not exist).

Note that successful response from manifest DELETE method is `202 Accepted` which suggests that deletion will be performed asynchronously, so I retry some time later expecting to get `404 Not Found`. I think `201 No Content` would be more appropriate when operation in synchronous.

stevvooe commented:
@bartebor Check out #2520 then.

bartebor commented:
@stevvooe #2520 is definitely a step in good direction, because it makes possible to delete other tags.
Unfortunately, it does not heal the repository and such broken tag becomes undeletable:
```
$ docker push devhub:5100/test:test1
The push refers to a repository [devhub:5100/test]
fb502e56d48d: Pushed 
test1: digest: sha256:a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e size: 523
$ docker push devhub:5100/test:test2
The push refers to a repository [devhub:5100/test]
56382659a064: Pushed 
test2: digest: sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79 size: 523

# now delete file repositories/test/_manifests/tags/test1/current from repository to trigger error

$ curl -XDELETE http://devhub:5100/v2/test/manifests/sha256:a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
$ curl -XDELETE http://devhub:5100/v2/test/manifests/sha256:f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
$ curl http://devhub:5100/v2/test/tags/list
{"name":"test","tags":["test1"]}

$ tree repositories/test/_manifests/
├── revisions
│   └── sha256
│       ├── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
│       └── f7a2a943efa23b421fc346c735d11e5229078b8673815addb0fe123ed905df79
└── tags
    └── test1
        └── index
            └── sha256
                └── a5baf6e65d49e036a2d31310a807ff0aa679021100c2e54ccc4265c78dd4f62e
                    └── link
```
GC will remove blobs, but `tag1` remains. It would be nice to have such tags removed eventually, but it is something one can live with.

bartebor commented:
@stevvooe I have encountered another `current/link` missing error. I'm currently running registry with #2520, so apart from one broken tag the repository runs ok. This time it happened during tag deletion when swift responded with error:
```
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.329216971+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_148f5f92-93ff-4088-b33b-c9cc68f32d77/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=8.878973ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=95f52025-c771-4bfc-af8c-a4d4ed56a303 trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.339418536+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_1bf1f544-9141-4c94-b32e-5653e06d59b0/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=10.076472ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=31a19a8d-3e8f-4d90-b557-942f07e31537 trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.350042407+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_37819c1f-4c0a-4491-b1d9-e11463b31972/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=10.477409ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=a02e6649-4349-4ee0-88bc-b5d6cb7c4902 trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.358547325+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_4797420d-607a-403b-9193-84671595c83e/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=8.330352ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=ed7f8315-3331-4f57-a8e7-a874f43e5f2b trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.37106869+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_484ed3a9-fb3e-4ea8-8b5d-0b70c52173c7/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=12.391891ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=db0742a6-20f9-44f6-aeb8-d97028d67136 trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.37879618+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_6d7a7574-a926-4b26-adf2-9b224a06a04f/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=7.617749ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=2b0d4413-817e-4bf6-92c0-467d7f05045a trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.386262684+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_a940b1ca-ce05-47a9-8f6a-1d419f5fe4bb/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=7.298472ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=575587bb-7f4b-45fe-bca5-8f29fe80aff1 trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:05 dockerhub registry[2018]: time="2018-02-19T05:06:05.393864636+01:00" level=debug msg="swift.GetContent("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/stress_c57d1c75-399b-4377-be19-8ab960edbfb5/current/link")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=7.467829ms trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=decaa842-03b3-43a1-af7d-cb6fc5334a1b trace.line=95 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:15 dockerhub registry[2018]: time="2018-02-19T05:06:15.441429525+01:00" level=debug msg="swift.Delete("/docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/overseer-stress_47ac42c7-29ed-43de-8276-310b4439a018")" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=10.047398869s trace.file="/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Delete" trace.id=2f8789df-8641-4b3d-bbeb-51102ac29630 trace.line=202 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
lut 19 05:06:15 dockerhub registry[2018]: time="2018-02-19T05:06:15.44173965+01:00" level=error msg="response completed with error" err.code=unknown err.message="swift: HTTP Error: 504: 504 Gateway Time-out" go.version=go1.8.1 http.request.host="dockerhub:443" http.request.id=f149ba4d-45db-48de-baef-69c1d94a72fa http.request.method=DELETE http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-stress/manifests/sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" http.response.contenttype="application/json; charset=utf-8" http.response.duration=10.341619216s http.response.status=500 http.response.written=167 vars.name="szwadron/overseer-stress" vars.reference="sha256:cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd"
```

On storage, there is only one file present:
```
docker/registry/v2/repositories/szwadron/overseer-stress/_manifests/tags/overseer-stress_47ac42c7-29ed-43de-8276-310b4439a018/index/sha256/cb11bb8f124a7a40ad73ef75cc17288acda915b2120f82495794d19ae80204bd/link
```

I think it would be nice to have storage inconsistencies collected automatically, because such tags are no longer deletable via registry API.

Vad1mo commented:
@bartebor Is this related to #1030 or to the PR #2441

bartebor commented:
@Vad1mo I think #1130 is not relevant. This thread is more about dealing with storage inconsistency should it happen.

I like #2441 concept a lot. This would probably eliminate the root cause of this problem altogether, not counting massive performance improvements with repository and tag operations.  I will definitely follow that PR, thanks for sharing :)

bartebor commented:
@stevvooe I catched an earlier mentioned error while tagging. PUT failed with 500 and left only one file:
```
[...]/_manifests/tags/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010/index/sha256/fa112f63bfe641e02240dc0ae24df0ddbf36b505867b6ae01b715254dc2a670d/link
```
Here are the logs:
```
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.033829081+01:00" level=debug msg="authorizing request" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.033991967+01:00" level=debug msg=PutImageManifest go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.034273325+01:00" level=debug msg="(*manifestStore).Put" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.03430738+01:00" level=debug msg="(*schema2ManifestHandler).Put" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.802687148+01:00" level=debug msg="swift.Stat("/docker/registry/v2/blobs/sha256/fa/fa112f63bfe641e02240dc0ae24df0ddbf36b505867b6ae01b715254dc2a670d/data")" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=768.075257ms trace.file="/tmp/docker_registry_rpmbuild/BUILD/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=cea53fea-4273-4aa6-9224-51dbf3047bcb trace.line=155 vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.818255922+01:00" level=debug msg="swift.PutContent("/docker/registry/v2/blobs/sha256/fa/fa112f63bfe641e02240dc0ae24df0ddbf36b505867b6ae01b715254dc2a670d/data")" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=15.451443ms trace.file="/tmp/docker_registry_rpmbuild/BUILD/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent" trace.id=f3ec87da-4de5-4e9f-aeca-873547e78ed6 trace.line=110 vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:55:56 dockerhub registry[2018]: time="2018-02-26T09:55:56.830056546+01:00" level=debug msg="swift.PutContent("/docker/registry/v2/repositories/szwadron/overseer-webserver/_manifests/revisions/sha256/fa112f63bfe641e02240dc0ae24df0ddbf36b505867b6ae01b715254dc2a670d/link")" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=11.681261ms trace.file="/tmp/docker_registry_rpmbuild/BUILD/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent" trace.id=48558219-3f6a-451a-b5c8-d91deb9b454e trace.line=110 vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:56:06 dockerhub registry[2018]: time="2018-02-26T09:56:06.836050143+01:00" level=debug msg="swift.PutContent("/docker/registry/v2/repositories/szwadron/overseer-webserver/_manifests/tags/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010/index/sha256/fa112f63bfe641e02240dc0ae24df0ddbf36b505867b6ae01b715254dc2a670d/link")" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" trace.duration=10.005492046s trace.file="/tmp/docker_registry_rpmbuild/BUILD/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent" trace.id=4332cd60-049d-411d-9647-860f32513c5f trace.line=110 vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
lut 26 09:56:06 dockerhub registry[2018]: time="2018-02-26T09:56:06.836277445+01:00" level=error msg="response completed with error" err.code=unknown err.detail="swift: HTTP Error: 504: 504 Gateway Time-out" err.message="unknown error" go.version=go1.8.1 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host="dockerhub:443" http.request.id=929c06a6-febd-4995-b3e4-de9645c434a2 http.request.method=PUT http.request.remoteaddr=10.32.33.19 http.request.uri="/v2/szwadron/overseer-webserver/manifests/overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010" http.request.useragent="Apache-HttpClient/4.3.6 (java 1.5)" http.response.contenttype="application/json; charset=utf-8" http.response.duration=10.804771525s http.response.status=500 http.response.written=167 vars.name="szwadron/overseer-webserver" vars.reference="overseer-webserver_c4a3144c-650c-49e2-a7df-b9a0e5cde010"
```

mrueg commented:
@bartebor can you fix the file conflict?

bartebor commented:
@mrueg Conflict emerged because of merging #2520 which addressed similar problem. I have agreed with @stevvooe that my version may be risky, so I would not merge this in. 

Storage corruption is still happening, but is not as destructive as before. We still need solution to this problem, so I did not close this PR. Feel free to do so if you think that such discussion should happen elsewhere, ie. #2430.
